src/ConnectionAutomaton.cpp

```diff
@@ -50,40 +48,41 @@ ConnectionAutomaton::~ConnectionAutomaton() {
   // We rely on SubscriptionPtr and SubscriberPtr to dispatch appropriate
   // terminal signals.
   DCHECK(!resumeCallback_);
+  DCHECK(
+      !frameTransport_); // the instance should be closed by via close() method
 }
 
 void ConnectionAutomaton::setResumable(bool resumable) {
-  DCHECK(!connection_); // we allow to set this flag before we are connected
+  DCHECK(!frameTransport_); // we allow to set this flag before we are connected
   isResumable_ = resumable;
 }
 
 void ConnectionAutomaton::connect(
-    std::unique_ptr<DuplexConnection> connection) {
-  CHECK(static_cast<bool>(connection_) ^ static_cast<bool>(connection));
-
-  if (connection) {
-    connection_ = std::move(connection);
-  }
-
-  connectionOutput_.reset(connection_->getOutput());
-  connectionOutput_.onSubscribe(shared_from_this());
-
-  // the onSubscribe call on the previous line may have called the terminating
-  // signal which would call disconnect/close
-  if (connection_) {
-    // This may call ::onSubscribe in-line, which calls ::request on the
-    // provided
-    // subscription, which might deliver frames in-line.
-    // it can also call onComplete which will call disconnect/close and reset
-    // the connection_ while still inside of the connection_::setInput method.
-    // We will create a hard reference for that case and keep the object alive
-    // until setInput method returns
-    auto connectionCopy = connection_;
-    connectionCopy->setInput(shared_from_this());
-  }
-
-  if (connection_) {
+    std::shared_ptr<FrameTransport> frameTransport) {
+  CHECK(!frameTransport_);
+  CHECK(frameTransport);
+  CHECK(!frameTransport->isClosed());
+
+  frameTransport_ = std::move(frameTransport);
+  // We need to create a hard reference to frameTransport_ to make sure the
+  // instance survives until the setFrameProcessor returns. There can be
+  // terminating signals processed in that call which will nullify
+  // frameTransport_
+  auto frameTransportCopy = frameTransport_;
+  frameTransport_->setFrameProcessor(shared_from_this());
+
+  // setFrameProcessor starts pulling frames from duplex connection
+  // the connection may also get closed before the method returns so
+  // we need to check again
+  auto outputFrames = streamState_->moveOutputFrames();
+  if (frameTransport_) {
+    for (auto& frame : outputFrames) {
+      outputFrame(std::move(frame));
+    }
     onConnected_();
+  } else {
+    LOG_IF(WARNING, !outputFrames.empty()) << "transport closed, throwing away "
```

Comment:
I mainly don't think you care in a server that is processing a huge volume of traffic with asynchronous network events and payload deliveries.  it's very inevitable and uninteresting.

VLOG?